docs/proposals/templates.md

```diff
@@ -0,0 +1,349 @@
+  the existing k8s codec scheme rather than having to define/introduce a new syntax evolution mechanism.
+  * (Even if they are not part of the k8s core, it would still be good if a server side template processing+managing api supplied 
+    as an ApiGroup consumed the same k8s object schema as the peer k8s apis rather than introducing a new one)
+  * This goal would be met by a solution which wraps some other template content (eg jinja, python, etc) inside a k8s object wrapper
+* Self-contained parameter definitions.  This allows a template to be a portable object which includes metadata that describe
+  the inputs it expects, making it easy to wrapper a user interface around the parameterization flow.
+* Support for non-string parameters.  K8S objects define fields of type string, int, bool, and byte[](base64).  The parameter
+  scheme should support all of those types for substitution.
+  * complex types (struct/map/list) can be defined in terms of the available primitives, so it's preferred to avoid the complexity
+    of allowing for full complex-type substitution.
+* Parameter metadata.  Parameters should include at a minimum, information describing the purpose of the parameter, whether it is 
+  required/optional, and a default/suggested value.  Type information could also be required to resolve the non-string substitution
+  requirement.
+* Template metadata.  Templates should be able to include metadata describing their purpose or links to further documentation and 
+  versioning information.
+* Pluggable template expansion.  It should be possible to define multiple template expansion strategies and the system should
+  delegate a template to the correct plugin for expansion and then return the result.
+
+## Proposed Object Syntax
+
+Example of a template definition:
+```
+kind: TemplateDeclaration
+apiVersion: v1
+metadata: 
+  name: myHelloWorldTemplate
+templateKind: python
+params:
+  - name: message
+    required: true
+    default_value: Hello World!
+    type: string
+  - name: port
+    required: true
+    default_value: 8080
+    type: int
+includes:
+  - path: https://github.com/myorg/mytemplates/template1.yml
+  - path: https://github.com/myorg/mytemplates/template2.yml
+body: |
+  import json
+  import kubernetes_tmpl
+  # Real templates would be doing actual computation.
+  print json.dumps({
+    'apiVersion': 'v1',
+    'kind': 'Pod',
+    'metadata': {
+      'name': kubernetes_tmpl.GetVar('message'),
+      'labels': {'name': 'nginx'},
+    },
+  })
+```
+
+Where "templateBody" is arbitrary content that will be processed by the appropriate template plugin, as selected by templateKind.
+
+Because the template body is arbitrary and various plugins can be provided, the overall capability and syntax of templates 
+is not dictated by this specification.
+
+The schema also provides for an array of references to additional templates that should be expanded when this template is
```

Comment:
I don't see this array of references in either example.  Is this about nesting templates?  We should have an example for that if so.